### PR TITLE
Add bits for BlobDataType in BlobId flag.

### LINKED ISSUE #940: Introduce a bit in BlobId to indicate simple/data/metadata blob
It is useful to know whether a blob is metadata blob in two known (and maybe more) situations
1) Background delete
2) TTL update

Having this information will help avoid an unnecessary GET

### PR DESCRIPTION
**Pull request by lightningrob**
Fixes #940.
Two bits in flag used for data type values {SIMPLE, COMPOSITE, DATACHUNK}.
At this stage, all non-composite blobs are tagged as DATACHUNK as I encountered issues distinguishing between SIMPLE and DATACHUNK.
Issue #971 was raised to track the remaining work.

### CODE REVIEW

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 1]
```diff
@@ -587,6 +674,26 @@ public static short getVersion(String idStr) throws IOException {
...
+  public enum BlobDataType {
+    /**
+     * Indicates a fully contained blob.
+     */
+    SIMPLE,
+
+    /**
+     * Indicates a composite (metadata) blob.
+     */
+    COMPOSITE,
```
Comment (zzmao) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: In most of the comments,  `COMPOSITE` is assumed the whole blob(meta + data). 
`METADATA` may be more appropriate here.

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 2]
```diff
@@ -880,14 +883,33 @@ private void prepareForBuilding(int chunkIndex, int size) {
      */
     private void prepareForSending() {
       try {
+        // Determine data type to set
+        BlobDataType blobDataType = null;
+        if (isMetadataChunk()) {
+          blobDataType = BlobDataType.COMPOSITE;
+        } else if (this.chunkIndex == 0) {
+          if (chunkFillingCompletedSuccessfully) {
+            // Note: this variable can be true even if additional chunks are filled!
```
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: Good point.

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 3]
```diff
@@ -587,6 +674,26 @@ public static short getVersion(String idStr) throws IOException {
     CRAFTED
   }
 
+  /**
+   * Indicates the type of blob this is.
+   */
+  public enum BlobDataType {
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: +1

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 1 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: This was copied from NotificationBlobType (as opposed to simply creating a member variable of that enum).  Shouldn't they be consistent?  Why do we use Composite in the existing enum?

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: That seems okay.

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 4]
```diff
@@ -99,6 +100,24 @@
  * |  un-assigned | IsEncrypted |  BlobIdType   |
  * +--------------+-------------+---------------+
  *
+ * </pre>
+ *
+ * <br>
+ * Version 5, which is the same as Version 4 but with additional info in the flag byte.  Flag bits 4 and 5
+ * are used to specify the Blob Data Type, which can be Simple, Composite, or DataChunk.
```
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: done

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Discussed with Gopal and Casey and decided to have an explicit new BlobId version.  Casey mentioned serialization issues and need to upgrade server prior to frontend.

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: Removed the inner if clause.

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 5]
```diff
@@ -141,7 +164,29 @@
    */
   public BlobId(short version, BlobIdType type, byte datacenterId, short accountId, short containerId,
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: this constructor should take in the `BlobDataType`. There should be no reason to construct a new blob id with a null value for the `BlobDataType` - correct? (or in other words, what piece of production code would use this constructor?)

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 6]
```diff
@@ -191,6 +237,14 @@ private BlobId(short version, BlobIdType type, byte datacenterId, short accountI
         this.containerId = containerId;
         this.isEncrypted = isEncrypted;
         break;
+      case BLOB_ID_V5:
+        this.type = type;
+        this.datacenterId = datacenterId;
+        this.accountId = accountId;
+        this.containerId = containerId;
+        this.isEncrypted = isEncrypted;
+        this.blobDataType = blobDataType;
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: should set this to null in the other cases

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 7]
```diff
@@ -382,6 +445,17 @@ protected String getUuid() {
         idBuf.putShort(accountId);
         idBuf.putShort(containerId);
         break;
+      case BLOB_ID_V5:
+        flag = (byte) (type.ordinal() & BLOB_ID_TYPE_MASK);
+        flag |= isEncrypted ? IS_ENCRYPTED_MASK : 0;
+        if (blobDataType != null) {
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: why would it be `null` in version 5?

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 8]
```diff
@@ -598,6 +705,7 @@ public static short getVersion(String idStr) throws IOException {
     final short accountId;
     final short containerId;
     final boolean isEncrypted;
+    BlobDataType blobDataType;
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: `final`

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 9]
```diff
@@ -608,6 +716,7 @@ public static short getVersion(String idStr) throws IOException {
     BlobIdPreamble(DataInputStream stream) throws IOException {
       version = stream.readShort();
       byte blobIdFlag;
+      blobDataType = null;
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: set this to null in the individual cases

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 10]
```diff
@@ -640,6 +749,16 @@ public static short getVersion(String idStr) throws IOException {
           accountId = stream.readShort();
           containerId = stream.readShort();
           break;
+        case BLOB_ID_V5:
+          blobIdFlag = stream.readByte();
+          type = BlobIdType.values()[blobIdFlag & BLOB_ID_TYPE_MASK];
+          isEncrypted = (blobIdFlag & IS_ENCRYPTED_MASK) != 0;
+          int dataTypeOrdinal = (blobIdFlag & BLOB_DATA_TYPE_MASK) >> BLOB_DATA_TYPE_SHIFT;
+          blobDataType = dataTypeOrdinal < BlobDataType.values().length ? BlobDataType.values()[dataTypeOrdinal] : null;
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: I don't think we should set it to `null` here based on the check. If the ser-deser etc went right, there should _never_ be a reason for `dataTypeOrdinal >= BlobDataType.values().length`. I'd rather see it fail if that happens (with the `IndexOutOBoundsException` that the enum will throw)

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 11]
```diff
@@ -562,6 +577,14 @@ private void assertBlobIdFieldValues(short version, BlobId blobId, BlobIdType ty
         assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
         assertEquals("Wrong isEncrypted value in blobId: " + blobId, isEncrypted, BlobId.isEncrypted(blobId.getID()));
         break;
+      case BLOB_ID_V5:
+        assertEquals("Wrong type in blobId: " + blobId, type, blobId.getType());
+        assertEquals("Wrong datacenter id in blobId: " + blobId, datacenterId, blobId.getDatacenterId());
+        assertEquals("Wrong account id in blobId: " + blobId, accountId, blobId.getAccountId());
+        assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
+        assertEquals("Wrong isEncrypted value in blobId: " + blobId, isEncrypted, BlobId.isEncrypted(blobId.getID()));
+        assertEquals("Wrong blobDataType value in blobId: " + blobId, blobDataType, blobId.getBlobDataType());
```
Comment (vgkholla) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: should probably check that this is null in the other versions?

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 12]
```diff
@@ -880,14 +883,27 @@ private void prepareForBuilding(int chunkIndex, int size) {
      */
     private void prepareForSending() {
       try {
+        // Determine data type to set
+        BlobDataType blobDataType = null;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: minor: initialization to `null` not required I think?

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 13]
```diff
@@ -880,14 +883,27 @@ private void prepareForBuilding(int chunkIndex, int size) {
...
+        if (isMetadataChunk()) {
+          blobDataType = BlobDataType.METADATA;
+        } else if (this.chunkIndex == 0) {
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: minor: qualifier `this.` not required?

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 5 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: I don't disagree, but this constructor is called in 63 places.  Should we change all of them?  It seems easier to maintain to use fluent constructors and setters, eg. new BlobId().version(5).type(NATIVE).etc()

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 7 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: It shouldn't, but I was getting NPE due a test not setting.  That test may be fixed now but it doesn't hurt to check for null.

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 8 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: done

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 9 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: done

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 10 above)
Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: ok

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 7 above)
Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: My preference is don't add extra check for `blobDataType `:
If not because of tests, might be a corner case that is not covered. Throwing NPE is helpful to debug. The similar variable is `isEncrypted `, no extra check is needed.
If it's because of tests, then must be something wrong in test. Let's fix it.

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java (hunk 11 above)
Comment (lightningrob) [reply ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: ok

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 5 above)
Comment (vgkholla) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Agreed on having a builder sort of construction on this. We should do this in a follow up PR.

**Comment by lightningrob** [comment]
@vgkholla Are the updated changes sufficient?

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 7 above)
Comment (vgkholla) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: +1 - no null checks here. It shouldn't _be_ `null` (i.e `blobDataType == null` is an `Exception` scenario)

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 14]
```diff
@@ -127,7 +129,7 @@ public void testBlobIdFlag() throws Exception {
       for (BlobIdType type : BlobIdType.values()) {
         for (boolean isEncrypted : isEncryptedValues) {
           BlobId blobId = new BlobId(version, type, referenceDatacenterId, referenceAccountId, referenceContainerId,
-              referencePartitionId, isEncrypted);
+              referencePartitionId, isEncrypted, null);
```
Comment (vgkholla) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: maybe this was the test that had a problem with NPE?

Comment (vgkholla) [reply ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: Couple of other places too

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java (hunk 7 above)
Comment (vgkholla) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Seems like doing this would cause you to change a lot of the tests - have to chat about how to make this work

Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: @vgkholla @zzmao I tried enforcing non-null, but this fails in the blobid crafting scenario which takes an existing blobId with version < 5 and up-versions it to the latest (5).  The original blobId has null data type.  How do you propose to handle this case?  I think we need to allow null value, and @cgtz seems to agree.

Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: If we allow null value, it may mask future bug. 
We can modify the test to do sth special for version<5.

Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: @zzmao It's not the test, it's the crafting functionality itself.  What do you suggest?

Comment (vgkholla) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: I don't think we need to worry about crafting too much since it is an offline operation performed with a specific goal in mind - which means that it is reasonable to expect that the crafting script has all the info required to construct a blob ID in the version required.

Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: @vgkholla but are they/we able to know `blobDataType` or `metaDataType`(v1-v4) before run crafting script? What if unknown?

Comment (vgkholla) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: then a v5 version of v1-v4 cannot be created. The blob must also be crafted in the same version.

Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Ah. I see. Then we are safe.

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 15]
```diff
@@ -224,6 +224,9 @@ private BlobId(short version, BlobIdType type, byte datacenterId, short accountI
         this.accountId = accountId;
         this.containerId = containerId;
         this.isEncrypted = isEncrypted;
+        if (blobDataType == null) {
+          throw new IllegalArgumentException("blobDataType can't be null for id version " + BLOB_ID_V5);
+        }
         this.blobDataType = blobDataType;
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: this is a tad better for readability
`this.blobDataType = Objects.requireNonNull(blobDataType, "blobDataType can't be null for id version " + BLOB_ID_V5);`

Comment (lightningrob) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Perhaps, but that utility throws NPE and IllegalArg is more precise IMO.  The same pattern is used for other parameters like partitionId.  I'd prefer to implement our own utility that throws the exception we want, and use that throughout.  I can take this in a follow up PR.
